### ISSUE TITLE
Rationale behind "RT" value in prio column

### ISSUE THREAD

**Issue by abusque** [hishamhm/htop#314]
Hi,

I've been working with task priorities on Linux quite a bit over the last few days, and I used htop to monitor my changes most of the time. During that time, I noticed that htop displays "RT" in the priority column only when the task priority reaches -100. See [`Process.c:427`](https://github.com/hishamhm/htop/blob/master/Process.c#L427). I was wondering what's the rationale behind this, given that, from the kernel's point of view at least, all negative task priorities correspond to real-time priorities. This can be seen at [`include/linux/sched/rt.h:6`](https://github.com/torvalds/linux/blob/master/include/linux/sched/rt.h#L6) in the kernel's source tree. Keep in mind that the priority seen by htop (or through `/proc`) is offset by `MAX_RT_PRIO`.

So, I can understand not wanting to print "RT" for all negative priority values, as this would hide too much information, but then, why put it only for a -100 task priority, instead of not at all?

**Comment by hishamhm** [comment]
> So, I can understand not wanting to print "RT" for all negative priority values, as this would hide too much information, but then, why put it only for a -100 task priority, instead of not at all?

This code is really old (I traced the "RT" line back to the initial import of this git repository, from back in 2006!) and I don't remember exactly what the rationale was, to be honest... I couldn't find anything in my email archives, either. I _think_ I just copied the behavior of the version of top that was current back in the day. It certainly might be worth revisiting.

**Comment by abusque** [comment]
That was my impression from a quick `git blame` as well. A quick look at procps-ng's top (which is the version currently shipped by Arch Linux, and seems to have superseded the original procps in most major distros) show similar behaviour. See [`top/top.c:5484`](https://gitlab.com/procps-ng/procps/blob/master/top/top.c#L5484), or [`top.c:3123`](http://procps.cvs.sourceforge.net/viewvc/procps/procps/top.c?view=markup) for the original procps.

The difference is that they display "RT" for all values below -99, not just -100, which I think correctly maps to task priorities from (much) older kernels.

I think the options would then be:

* Retain the current behaviour for "compatibility" with other top implementations, although I'm not sure how much of a factor this is for htop.
* Not displaying "RT" for any priority value, though this increases the user's cognitive load by requiring them to know that negative priorities are for real-time tasks.
* Displaying "RT" for all negative priority valuee. I think this option hides too much useful information.
* Displaying the "RT" status of a process separately from the numerical prio value. Whether it would be a different column, a different colour for the numerical prio value, or something else altogether.

If a change is to be made, I think I'd favour the last option, although I don't have a definitive solution right now as to how to best display this extra information.

Also, finally, I imagine we'd have to consider how task priorities and real-time tasks are handled on other operating systems, as I've only described the behaviour on Linux so far.

**Comment by hishamhm** [comment]
> That was my impression from a quick git blame as well. A quick look at procps-ng's top (which is the version currently shipped by Arch Linux, and seems to have superseded the original procps in most major distros) show similar behaviour. See top/top.c:5484, or top.c:3123 for the original procps.

Thank you for the detective work!

> Displaying the "RT" status of a process separately from the numerical prio value. Whether it would be a different column, a different colour for the numerical prio value, or something else altogether.

I like the latter option the best, too. I think a different color for RT would be best. So, what range of numbers actually correspond to RT priorities nowadays?

**Comment by hishamhm** [comment]
A small UI issue I'm going through here is that I don't really want to extend the PRI column another character just to fit "-100"...

**Comment by hishamhm** [comment]
Further looking into this:

> The difference is that they display "RT" for all values below -99, not just -100, which I think correctly maps to task priorities from (much) older kernels.

Isn't it the case that RT prios go from -1 to -100, so the only value below -99 is -100? It seems so from here: [`linux/sched/prio.h`](https://github.com/torvalds/linux/blob/master/include/linux/sched/prio.h)

I'm leaning towards painting all negative numbers red, and writing -100 as "max", indicating maximum priority.

**Comment by abusque** [comment]
Yeah I had thought  about the extra character as well, not sure how to best deal with this. I guess if horizontal real-estate is really scarce, it might be a factor in favour of retaining the current behaviour. I think the max option is a reasonable solution as well.

As for RT prio values in current Linux kernels, as seen by htop, the range is [-100..-1], and [0..39] for non-RT values (although I believe the value -1 can't happen normally happen in the current implementation, but it is not strictly an invalid priority value). So yes, all negative values are real time, and technically `prio  < -99` is only going to happen for `prio == -100`, under the current implementation.

As a more technical aside, this value is fetched via the `task_prio` function in the kernel at [`kernel/sched/core.c:3505`](http://lxr.free-electrons.com/source/kernel/sched/core.c#L3505). The comment above the function is outdated (it predates the git import of the kernel source from back in 2005) and describes the older behaviour, which could lead, amongst other things, to possible return values below -99. I've submitted [a patch](https://lkml.org/lkml/2015/11/28/158) to update that comment to the LKML. Basically, instead of exposing the `[0..MAX_RT_PRIO - 1] `and `[MAX_RT_PRIO..MAX_PRIO - 1]` ranges internal to  the kernel, it offsets by `MAX_RT_PRIO`, which is why `/proc`, and htop, see negative values, in thje range `[-MAX_RT_PRIO..-1] ` for RT tasks and `[0..MAX_PRIO - MAX_RT_PRIO - 1]` for normal tasks.

**Comment by hishamhm** [comment]
One thing that got me confused as I read about this (following Stack Overflow posts, etc.) was the exact meaning of the negative numbers: is -100 the "top priority" (runs ahead of everything else) or the "least priority among the RT processes (runs after all other RT processes ran, but ahead of the non-RT processes)"?

**Comment by abusque** [comment]
-100 is the highest priority. It's quite confusing, which is why I spent a bit of time fiddling around with this and looking at the results in different places in `/proc` and htop. Outdated docs in the kernel just led to more confusion.

If you want you can have a look at the `chrt` utility, which helped clear things up for me. You can set a priority from 1 to 99 for a real-time task, where 1 is truly the lowest RT priority, and 99 the highest. If you `chrt -p 1 <pid>` on a RT pid, and then look at the result in htop, you'll see -2 as the prio value, and for  `chrt -p 99 <pid>`, 'RT', or -100. Positive prio values for non-RT processes just map to their nice value, [0..39] prio corresponds to [-20..19] nice.
